SpatialGDK/Source/SpatialGDK/Private/Interop/Startup/ElectGsmAuthWorkerStep.cpp

```diff
@@ -0,0 +1,121 @@
+#include "Interop/Startup/ElectGsmAuthWorkerStep.h"
+
+#include "Algo/MinElement.h"
+#include "Interop/GlobalStateManager.h"
+#include "Interop/Startup/ServerWorkerStartupCommon.h"
+
+namespace SpatialGDK
+{
+class FTryClaimingGsmPartitionStep : public FStartupStep
+{
+public:
+	FTryClaimingGsmPartitionStep(TSharedRef<FInternalState> InState, UGlobalStateManager& InGlobalStateManager,
+								 TSharedRef<TOptional<bool>> bInShouldHaveGsmAuthority)
+		: State(InState)
+		, GlobalStateManager(&InGlobalStateManager)
+		, bShouldHaveGsmAuthority(bInShouldHaveGsmAuthority)
+	{
+		check(!bShouldHaveGsmAuthority->IsSet());
+
+		StepName = TEXT("Claiming GSM partition");
+	}
+
+	virtual bool TryFinish() override;
+
+	bool TryClaimingPartition();
+
+private:
+	TSharedRef<FInternalState> State;
+	UGlobalStateManager* GlobalStateManager;
+	TSharedRef<TOptional<bool>> bShouldHaveGsmAuthority;
+};
+
+class FWaitForGsmAuthorityStep : public FStartupStep
+{
+public:
+	FWaitForGsmAuthorityStep(TSharedRef<FInternalState> InState, UGlobalStateManager& InGlobalStateManager,
+							 TSharedRef<TOptional<bool>> bInShouldHaveGsmAuthority)
+		: State(InState)
+		, GlobalStateManager(&InGlobalStateManager)
+		, bShouldHaveGsmAuthority(bInShouldHaveGsmAuthority)
+	{
+		StepName = TEXT("Waiting to receive GSM authority");
+	}
+
+	virtual bool TryFinish() override;
+
+private:
+	TSharedRef<FInternalState> State;
+	UGlobalStateManager* GlobalStateManager;
+	TSharedRef<TOptional<bool>> bShouldHaveGsmAuthority;
+};
+
+FElectGsmAuthWorkerStep::FElectGsmAuthWorkerStep(TSharedRef<FInternalState> InState, UGlobalStateManager& InGlobalStateManager)
+	: State(InState)
+	, GlobalStateManager(&InGlobalStateManager)
+	, Executor(CreateSteps())
+{
+}
+
+TArray<TUniquePtr<FStartupStep>> FElectGsmAuthWorkerStep::CreateSteps()
+{
+	TArray<TUniquePtr<FStartupStep>> Steps;
+	Steps.Emplace(MakeUnique<FTryClaimingGsmPartitionStep>(State, *GlobalStateManager, bShouldHaveGsmAuthority));
+	Steps.Emplace(MakeUnique<FWaitForGsmAuthorityStep>(State, *GlobalStateManager, bShouldHaveGsmAuthority));
+	return Steps;
+}
+
+bool FElectGsmAuthWorkerStep::TryFinish()
+{
+	return Executor.TryFinish();
+}
+
+bool FTryClaimingGsmPartitionStep::TryFinish()
```

Comment:
This TryFinish does not do any polling and always completes. Is it worth its own step? It looks like it could be in the Start method instead.

Reply:
Nice catch!